### ISSUE TITLE
Fast path for CommonJS on Module._compile

### ISSUE THREAD

**Issue by H4ad** [nodejs/performance#139]
In theory, we can move the [load of content of a module](https://github.com/nodejs/node/blob/main/lib/internal/modules/cjs/loader.js#L1396) to C++ when:

- the experimental policy is not enabled
- and when the module is not ESM
- and when the module is not patched

Since the `content` goes directly to the C++ without modification:

- https://github.com/nodejs/node/blob/main/lib/internal/modules/cjs/loader.js#L1288
- https://github.com/nodejs/node/blob/main/lib/internal/vm.js#L78
- https://github.com/nodejs/node/blob/95b198911382b0aa0a0448e2a86214558665e02b/src/node_contextify.cc#L975

---

Also, I don't have any idea if worth or is viable but maybe we could use [StreamedSource](https://github.com/v8/v8/blob/7fdc2728f2114b083ebbd24710a32fe4574c0c57/include/v8-script.h#L490-L496C29) instead of `Source`, in this way, we probably can save some memory and load the script much faster.

Let me know what you think, this is something that was in my mind for a while and I just wanna to share it with more people to understand if is possible or not.

**Comment by Uzlopak** [comment]
Pinging @nodejs/loaders

**Comment by GeoffreyBooth** [comment]
In general we've been treating the CommonJS loader as legacy that is frozen and not to be updated, because we don't have the resources to debug any issues that may result to changes there. No one on the current loaders/modules team has worked with the CommonJS loader, as it hasn't had updates (other than parts shared with the ESM loader) in years.

The ESM loader is much more ripe for optimizations, if I can encourage you to focus there.

**Comment by joyeecheung** [comment]
I don't think we've ever agreed that CommonJS loader is a legacy. It is frozen, but only because of the impact caused by compatibility. Many CLI tools (e.g. eslint, npm and tsc) still choose to bundle their sources as a CommonJS bundle to speed up start up time, so optimizations of the CommonJS loader would benefit the ecosystem through these popular apps that still use CommonJS.

> No one on the current loaders/modules team has worked with the CommonJS loader, as it hasn't had updates (other than parts shared with the ESM loader) in years.

I don't think this is true, looking at the git log of `lib/internal/modules/cjs/` I see >10 updates for SEA, policy and the permission model in just the past year, all from people outside of the `loaders/modules` team - those teams were specifically for ESM, whereas the CommonJS loader is a long-time resident that's much easier to hack on for contributors not familiar with any loader due to its smaller size and complexity.

**Comment by arcanis** [comment]
> In theory, we can move the [load of content of a module](https://github.com/nodejs/node/blob/main/lib/internal/modules/cjs/loader.js#L1396) to C++ when:

- the `fs` module isn't patched (virtual filesystems)

**Comment by H4ad** [comment]
@arcanis You mean if we move the loading of the content to the `C++`, we will not be able to patch `fs` to support virtual filesystems?

If this is true, I think we already broke the support for virtual filesystems by moving the `package reader` to C++ (during module resolution if I remember correctly).

**Comment by arcanis** [comment]
> If this is true, I think we already broke the support for virtual filesystems by moving the `package reader` to C++ (during module resolution if I remember correctly).

Yes, and it had to be [fixed](https://github.com/nodejs/node/pull/44537).

It's fine to use faster primitives, but there should be a way for virtual filesystem to still affect them. Ideally that would be through a blessed API, perhaps similar to the loader hooks, and failing that through discouraged access points.
